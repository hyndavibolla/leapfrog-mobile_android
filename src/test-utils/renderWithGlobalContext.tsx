import React, { ReactElement } from 'react';
import { render } from '@testing-library/react-native';

import { getInitialState, GlobalProvider } from '../state-mgmt/GlobalState';
import { Deps, IGlobalState, IReducer } from '../models/general';
import { getMockDeps } from './getMockDeps';
import { getMockReducer } from './getMockReducer';

export const renderWithGlobalContext = (
  children: ReactElement,
  deps: Deps = getMockDeps(),
  initState: IGlobalState = getInitialState(),
  combinedReducers: IReducer<IGlobalState, any> = getMockReducer()
) => {
  const renderAPI = render(
    <GlobalProvider deps={deps} initState={initState} combinedReducers={combinedReducers}>
      {children}
    </GlobalProvider>
  );
  return { ...renderAPI, mockReducer: combinedReducers, deps };
};
