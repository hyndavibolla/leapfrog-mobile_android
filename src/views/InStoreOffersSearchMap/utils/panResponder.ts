const swipeConfig = {
  velocityThreshold: 0.3,
  directionalOffsetThreshold: 80,
  gestureIsClickThreshold: 5
};

const swipeDirections = {
  swipeUp: 'SWIPE_UP',
  swipeDown: 'SWIPE_DOWN'
};

export const handleShouldSetPanResponder = ({ nativeEvent: { touches } }, gestureState) => {
  return touches.length === 1 && !gestureIsClick(gestureState);
};

export const gestureIsClick = ({ dx, dy }) => {
  return Math.abs(dx) < swipeConfig.gestureIsClickThreshold && Math.abs(dy) < swipeConfig.gestureIsClickThreshold;
};

export const handlePanResponderEnd = (gestureState, onSwipeUp, onSwipeDown) => {
  const swipeDirection = getSwipeDirection(gestureState);
  triggerSwipeHandlers(swipeDirection, onSwipeUp, onSwipeDown);
};

export const triggerSwipeHandlers = (swipeDirection, onSwipeUp, onSwipeDown) => {
  const { swipeUp, swipeDown } = swipeDirections;
  switch (swipeDirection) {
    case swipeUp:
      onSwipeUp();
      break;
    case swipeDown:
      onSwipeDown();
      break;
  }
};

export const getSwipeDirection = gestureState => {
  const { swipeUp, swipeDown } = swipeDirections;
  if (isValidVerticalSwipe(gestureState)) {
    return gestureState.dy > 0 ? swipeDown : swipeUp;
  }
  return null;
};

export const isValidVerticalSwipe = ({ vy, dx }) => {
  const { velocityThreshold, directionalOffsetThreshold } = swipeConfig;
  return isValidSwipe(vy, velocityThreshold, dx, directionalOffsetThreshold);
};

export const isValidSwipe = (velocity, velocityThreshold, directionalOffset, directionalOffsetThreshold) => {
  return Math.abs(velocity) > velocityThreshold && Math.abs(directionalOffset) < directionalOffsetThreshold;
};
