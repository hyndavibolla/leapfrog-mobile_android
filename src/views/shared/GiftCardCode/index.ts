export { default as GiftCardCode } from './GiftCardCode';
