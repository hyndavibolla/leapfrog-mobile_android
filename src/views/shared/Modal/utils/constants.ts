export enum ModalType {
  CENTER = 'center',
  BOTTOM = 'bottom'
}

export enum ModalSize {
  SMALL = 'small',
  MEDIUM = 'medium',
  EXTRA_LARGE = 'extralarge',
  DYNAMIC = 'dynamic',
  FULL_SCREEN = 'fullscreen'
}

export const ModalSizeToHeightMap = {
  [ModalSize.SMALL]: 40,
  [ModalSize.MEDIUM]: 50,
  [ModalSize.EXTRA_LARGE]: 80,
  [ModalSize.FULL_SCREEN]: 100
};
