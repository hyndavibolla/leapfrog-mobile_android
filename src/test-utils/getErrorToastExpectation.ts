import { act } from 'react-test-renderer';

import { ACTION_TYPE } from '../state-mgmt/core/actions';
import { ToastType } from '../views/shared/Toast';
import { Deps } from '../models/general';
import { wait } from '../utils/wait';
import { renderWrappedHook } from './renderWrappedHook';

export const getErrorToastExpectation = async (hookGetter: () => any, deps: Deps, errorArgs: any[] = []) => {
  const { result, mockReducer } = renderWrappedHook(hookGetter, deps);
  await act(async () => {
    await (result.current[0] as (...args: any[]) => Promise<any>)(...errorArgs);
    await wait(0); // waiting for the useEffect dispatching the toast to run
    expect(mockReducer).toBeCalledWith(
      expect.any(Object),
      expect.objectContaining({
        type: ACTION_TYPE.ADD_TOAST,
        payload: expect.objectContaining({ props: expect.objectContaining({ type: ToastType.ERROR }) })
      })
    );
  });
};
