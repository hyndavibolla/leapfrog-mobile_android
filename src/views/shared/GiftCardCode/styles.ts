import { css } from '@emotion/native';
import { COLOR, FONT_FAMILY, FONT_SIZE } from '../../../constants';

export const styles = {
  container: css`
    align-items: center;
    justify-content: center;
  `,
  fallbackContainer: css`
    align-items: center;
    flex-direction: column;
  `,
  fallbackTitle: css`
    margin-top: 20px;
  `,
  fallbackSubtitle: css`
    color: ${COLOR.DARK_GRAY};
    font-family: ${FONT_FAMILY.MEDIUM};
    font-size: ${FONT_SIZE.SMALLER};
    line-height: ${FONT_SIZE.HUGE};
    margin-top: 16px;
    text-align: center;
  `
};
