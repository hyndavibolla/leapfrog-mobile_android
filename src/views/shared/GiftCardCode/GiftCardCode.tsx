import React, { memo, useMemo } from 'react';
import { View, Text, StyleProp, ViewStyle, Dimensions } from 'react-native';
import LottieView from 'lottie-react-native';

import { useTestingHelper } from '../../../utils/useTestingHelper';
import { ImageWithFallbackAndLoading } from '../ImageWithFallbackAndLoading';
import GiftCardCodeFallback from '../../../assets/shared/giftCardCodeFallback.svg';
import { Title } from '../Title';
import { ENV } from '../../../constants';

import { styles } from './styles';

const windowWidth = Dimensions.get('window').width;

export enum CodeKind {
  CODE128 = 'CODE128',
  PDF417 = 'PDF417',
  DATA_MATRIX = 'DATA_MATRIX',
  QR_CODE = 'QR_CODE',
  PDF417_COMPACT = 'PDF417_COMPACT'
}

export interface Props {
  barcodeKind?: string;
  barcodeValue: string;
  imageHeight?: number;
  imageWidth?: number;
  colorHex?: string;
  fontColorHex?: string;
  fontSize?: number;
  loadingSize?: number;
  style?: StyleProp<ViewStyle>;
}

export const GiftCardCode = ({
  barcodeKind,
  barcodeValue,
  imageWidth,
  imageHeight,
  colorHex = 'FFFFFF',
  fontColorHex = '000000',
  fontSize = 0,
  loadingSize = 35,
  style
}: Props) => {
  const { getTestIdProps } = useTestingHelper('card-code');
  if (!imageWidth) {
    switch (barcodeKind?.toUpperCase()) {
      case CodeKind.CODE128:
      case CodeKind.PDF417:
      case CodeKind.PDF417_COMPACT:
        imageWidth = windowWidth * 0.8;
        imageHeight = 93;
        break;
      default:
        imageWidth = 170;
        imageHeight = 170;
    }
  }

  const source = useMemo(
    () => ({
      uri: `${ENV.TELLURIDE_API.CONTENT_BASE_URL}tellurideAS/PersonalizedEmail/genbc?msg=${barcodeValue}&height=${Math.round(imageHeight)}&width=${Math.round(
        imageWidth
      )}&fontSize=${fontSize}&bgColorHex=${colorHex}&fontColorHex=${fontColorHex}&barcode.kind=${barcodeKind}`
    }),
    [barcodeKind, barcodeValue, colorHex, fontColorHex, fontSize, imageHeight, imageWidth]
  );

  const Fallback = useMemo(
    () => (
      <View style={styles.fallbackContainer} {...getTestIdProps('fallback')}>
        <GiftCardCodeFallback />
        <Title style={styles.fallbackTitle}>We can't load this code right now</Title>
        <Text style={styles.fallbackSubtitle}>Please try again later.</Text>
      </View>
    ),
    [getTestIdProps]
  );

  return (
    <View style={[styles.container, style]}>
      <ImageWithFallbackAndLoading
        source={source}
        customFallback={Fallback}
        loading={
          <LottieView style={{ width: loadingSize, height: loadingSize }} loop autoPlay source={require('../../../assets/spinner/spinner-loader.json')} />
        }
        resizeMode="cover"
        style={{ width: imageWidth, height: imageHeight }}
        {...getTestIdProps('image')}
      />
    </View>
  );
};

export default memo(GiftCardCode);
