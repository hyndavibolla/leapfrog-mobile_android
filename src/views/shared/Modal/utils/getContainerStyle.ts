import { memorize } from '../../../../utils/memorize';
import { styles } from '../styles';
import { ModalSize, ModalType, ModalSizeToHeightMap } from './constants';

interface props {
  height: number;
  modalHeight: number;
  size: ModalSize;
  type: ModalType;
}

const getContainerStyle = ({ height, modalHeight, size, type }: props) => {
  switch (type) {
    case ModalType.BOTTOM:
      if (size === ModalSize.FULL_SCREEN) return [styles.containerBottom, styles.containerBottomFullScreen];
      else if (size !== ModalSize.DYNAMIC) return [styles.containerBottom, { maxHeight: (ModalSizeToHeightMap[size] * height) / 100 }];
      return [styles.containerBottom];
    case ModalType.CENTER:
      if (size === ModalSize.DYNAMIC) return [styles.containerCenter, styles.containerCenterDynamic, { bottom: (height - modalHeight) / 2 }];
      const center = (ModalSizeToHeightMap[size] * height) / 100;
      return [styles.containerCenter, { height: center, bottom: (height - center) / 2 }];
  }
};

export default memorize(getContainerStyle);
