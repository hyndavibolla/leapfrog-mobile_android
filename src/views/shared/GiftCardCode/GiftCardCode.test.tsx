import React from 'react';
import { act } from 'react-test-renderer';
import { renderWithGlobalContext } from '../../../test-utils/renderWithGlobalContext';
import { wait } from '../../../utils/wait';

import GiftCardCode, { Props, CodeKind } from './GiftCardCode';

describe('GiftCardCode', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      barcodeKind: CodeKind.CODE128,
      barcodeValue: '1234567890'
    };
  });

  it('should render', () => {
    const { toJSON } = renderWithGlobalContext(<GiftCardCode {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render correctly with size images', async () => {
    props.imageWidth = 200;
    props.imageHeight = 200;
    const { getByTestId } = renderWithGlobalContext(<GiftCardCode {...props} />);
    await act(() => wait(0));
    expect(getByTestId('image-with-fallback-container')).toBeTruthy();
  });

  it('should render correctly the images code', async () => {
    const { getByTestId } = renderWithGlobalContext(<GiftCardCode {...props} />);
    await act(() => wait(0));
    expect(getByTestId('image-with-fallback-container')).toBeTruthy();
  });

  it('should render correctly the images code with PDF417', async () => {
    props.barcodeKind = CodeKind.PDF417;
    const { getByTestId } = renderWithGlobalContext(<GiftCardCode {...props} />);
    await act(() => wait(0));
    expect(getByTestId('image-with-fallback-container')).toBeTruthy();
  });

  it('should render correctly the images code with PDF417_COMPACT', async () => {
    props.barcodeKind = CodeKind.PDF417_COMPACT;
    const { getByTestId } = renderWithGlobalContext(<GiftCardCode {...props} />);
    await act(() => wait(0));
    expect(getByTestId('image-with-fallback-container')).toBeTruthy();
  });

  it('should render correctly the images code with QR_CODE', async () => {
    props.barcodeKind = CodeKind.QR_CODE;
    const { getByTestId } = renderWithGlobalContext(<GiftCardCode {...props} />);
    await act(() => wait(0));
    expect(getByTestId('image-with-fallback-container')).toBeTruthy();
  });

  it('should render correctly the images code with DATA_MATRIX', async () => {
    props.barcodeKind = CodeKind.DATA_MATRIX;
    const { getByTestId } = renderWithGlobalContext(<GiftCardCode {...props} />);
    await act(() => wait(0));
    expect(getByTestId('image-with-fallback-container')).toBeTruthy();
  });
});
