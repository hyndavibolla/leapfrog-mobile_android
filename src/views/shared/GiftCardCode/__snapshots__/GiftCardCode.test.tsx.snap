// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiftCardCode should render 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "height": 93,
          "width": 600,
        },
      ]
    }
  >
    <FastImageView
      accessibilityLabel="image-with-fallback-container"
      accessible={true}
      onFastImageError={[Function]}
      onFastImageLoadEnd={[Function]}
      onFastImageLoadStart={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://content.uat.telluride.transformco.com/tellurideAS/PersonalizedEmail/genbc?msg=1234567890&height=93&width=600&fontSize=0&bgColorHex=FFFFFF&fontColorHex=000000&barcode.kind=CODE128",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="image-with-fallback-container"
    />
  </View>
</View>
`;
